frontend/src/pages/ExperimentDetails.test.tsx

```diff
@@ -0,0 +1,453 @@
+      match: { params: { eid: MOCK_EXPERIMENT.id } } as any,
+      toolbarProps: ExperimentDetails.prototype.getInitialToolbarState(),
+      updateBanner: updateBannerSpy,
+      updateDialog: updateDialogSpy,
+      updateSnackbar: updateSnackbarSpy,
+      updateToolbar: updateToolbarSpy,
+    };
+  }
+
+  async function mockNJobs(n: number): Promise<void> {
+    listJobsSpy.mockImplementation(() => ({
+      jobs: range(n).map(i => ({ id: 'test-job-id' + i, enabled: true, name: 'test job name' + i })),
+    }));
+    await listJobsSpy;
+    await TestUtils.flushPromises();
+  }
+
+  async function mockNRuns(n: number): Promise<void> {
+    listRunsSpy.mockImplementation(() => ({
+      runs: range(n).map(i => ({ id: 'test-run-id' + i, name: 'test run name' + i })),
+    }));
+    await listRunsSpy;
+    await TestUtils.flushPromises();
+  }
+
+  beforeEach(async () => {
+    // Reset mocks
+    consoleLogSpy.mockReset();
+    consoleErrorSpy.mockReset();
+    updateBannerSpy.mockReset();
+    updateDialogSpy.mockReset();
+    updateSnackbarSpy.mockReset();
+    updateToolbarSpy.mockReset();
+    getExperimentSpy.mockReset();
+    historyPushSpy.mockReset();
+    listJobsSpy.mockReset();
+    listRunsSpy.mockReset();
+
+    getExperimentSpy.mockImplementation(() => newMockExperiment());
+
+    await mockNJobs(0);
+    await mockNRuns(0);
+  });
+
+  it('renders a page with no runs or recurring runs', async () => {
+    const tree = shallow(<ExperimentDetails {...generateProps()} />);
+    await TestUtils.flushPromises();
+    expect(tree).toMatchSnapshot();
+    tree.unmount();
+  });
+
+  it('uses the experiment ID in props as the page title if the experiment has no name', async () => {
+    const experiment = newMockExperiment();
+    experiment.name = '';
+
+    const props = generateProps();
+    props.match = { params: { eid: 'test exp ID' } } as any;
+
+    getExperimentSpy.mockImplementationOnce(() => experiment);
+
+    const tree = shallow(<ExperimentDetails {...props} />);
+    await TestUtils.flushPromises();
+    expect(updateToolbarSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      pageTitle: 'test exp ID',
+      pageTitleTooltip: 'test exp ID'
+    }));
+    tree.unmount();
+  });
+
+  it('uses the experiment name as the page title', async () => {
+    const experiment = newMockExperiment();
+    experiment.name = 'A Test Experiment';
+
+    getExperimentSpy.mockImplementationOnce(() => experiment);
+
+    const tree = shallow(<ExperimentDetails {...generateProps()} />);
+    await TestUtils.flushPromises();
+    expect(updateToolbarSpy).toHaveBeenLastCalledWith(expect.objectContaining({
+      pageTitle: 'A Test Experiment',
+      pageTitleTooltip: 'A Test Experiment'
+    }));
+    tree.unmount();
+  });
+
+  it('uses an empty string if the experiment has no description', async () => {
+    const experiment = newMockExperiment();
+    experiment.description = undefined;
```

Comment:
nit: I think what you want here is `delete experiment.description`.

Reply:
Done.

just to avoid `= undefined` ?

Reply:
This is very nit-picky in this context, but it's a technical difference. If you set the field to undefined, it still exists on the object, and things like `expect(x).not.toHaveProperty` for example will fail. If we have logic that does something like `if 'description' in experiment` or `if experiment.hasOwnProperty(description)`, the test won't be testing that.
Again, very nit-picky. :)